### PR TITLE
[FLINK-35822] Introduce DESCRIBE FUNCTION

### PR DESCRIPTION
**Pull request by nateab**
## What is the purpose of the change

This PR introduces `DESCRIBE FUNCTION` syntax, allowing users to describe the metadata of a given function. The design for this is covered under FLIP-465: https://cwiki.apache.org/confluence/display/FLINK/FLIP-465%3A+Introduce+DESCRIBE+FUNCTION

## Brief change log
  - Add `Parser` support for the next syntax
  - Add `FunctionDefinition` as a new optional parameter for `ContextResolvedFunction`
  - The bulk of the implementation logic is in `DescribeFunctionOperation`

## Verifying this change

This change added tests and can be verified as follows:
  - Added end to end tests in `flink-table/flink-sql-client/src/test/resources/sql/function.q`
  - Added unit tests in `FlinkSqlParserImplTest` and `SqlOtherOperationConverterTest`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs / JavaDocs

### CODE REVIEW

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        if (!functionOpt.isPresent()) {
+            throw new ValidationException(
+                    String.format(
+                            "Function with the identifier '%s' doesn't exist.",
+                            sqlIdentifier.asSummaryString()));
+        }
+        ContextResolvedFunction function = functionOpt.get();
+        CatalogFunction catalogFunction = function.getCatalogFunction();
+
+        if (catalogFunction == null && !isExtended) {
```
Comment (AlanConfluent) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:89]: Not sure if it would be useful, but you could just output "System function": true if it's a system function, regardless, rather than throw an error.  Could be useful for seeing which function a name is bound to.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            rows.add(Arrays.asList("temporary", function.isTemporary()));
+        }
+
+        if (isExtended) {
+            FunctionDefinition definition = function.getDefinition();
+            rows.add(Arrays.asList("kind", definition.getKind().toString()));
+            rows.add(Arrays.asList("requirements", definition.getRequirements().toString()));
+            rows.add(Arrays.asList("deterministic", definition.isDeterministic()));
+            rows.add(Arrays.asList("constant folding", definition.supportsConstantFolding()));
+            rows.add(
```
Comment (AlanConfluent) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:114]: If we wanted to extend this to include return types, would that just be added to the signature I assume?  I don't think the format of these signatures is well defined, so maybe you're free to do that in the future.

File: flink-table/flink-sql-client/src/test/resources/sql/function.q [hunk 3]
```diff
@@ -346,3 +346,69 @@ show user functions;
...
+# test describe function
+# ==========================================================================
+
+ADD JAR '$VAR_UDF_JAR_PATH';
+[INFO] Execute statement succeeded.
+!info
+
+describe function temp_upperudf;
++-------------------+---------------------------------------------$VAR_UDF_JAR_PATH_DASH+
+|         info name |                                 $VAR_UDF_JAR_PATH_SPACE info value |
```
Comment (AlanConfluent) [inline flink-table/flink-sql-client/src/test/resources/sql/function.q:360]: What are these variables? `VAR_UDF_JAR_PATH_SPACE`. Seems some of the docs have them, and others don't but makes it hard to read the source file.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+                            "signature",
+                            generateSignature(
+                                    definition.getTypeInference(
+                                            ctx.getCatalogManager().getDataTypeFactory()),
+                                    function.toString(),
+                                    definition)));
+        }
+
+        return buildTableResult(
+                new String[] {"info name", "info value"},
```
Comment (AlanConfluent) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:125]: Nit: Maybe the columns can be just `name`, `value`? Info sounds a bit informal, but `metadata` sounds a bit much too.

File: flink-table/flink-sql-client/src/test/resources/sql/function.q [hunk 5]
```diff
@@ -1,4 +1,4 @@
-# function.q - CREATE/DROP/ALTER FUNCTION
+# function.q - CREATE/DROP/ALTER/SHOW/DESCRIBE FUNCTION
```
Comment (pgaref) [inline flink-table/flink-sql-client/src/test/resources/sql/function.q:1]:
```suggestion
# function.q - CREATE/DROP/ALTER/DESCRIBE FUNCTION
```

File: flink-table/flink-sql-client/src/test/resources/sql/function.q [hunk 6]
```diff
@@ -265,7 +265,7 @@ show user functions;
 # test alter function
 # ==========================================================================
 
-alter function func11 as 'org.apache.flink.table.client.gateway.local.ExecutorImplITCase$TestScalaFunction';
+alter function func11 as 'org.apache.flink.table.client.gateway.ExecutorImplITCase$TestScalaFunction';
```
Comment (pgaref) [inline flink-table/flink-sql-client/src/test/resources/sql/function.q:268]: Unrelated change?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeFunction.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    public List<SqlNode> getOperandList() {
+        return Collections.singletonList(functionNameIdentifier);
+    }
+
+    public boolean isExtended() {
+        return isExtended;
+    }
+
+    public String[] fullFunctionName() {
+        return functionNameIdentifier.names.toArray(new String[0]);
```
Comment (pgaref) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeFunction.java:68]:
```suggestion
        return functionNameIdentifier.names.toArray(new String[functionNameIdentifier.names.size]);
```

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java (hunk 2 above)
Comment (pgaref) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:114]: I understand the FLIP does not cover that, but why wont we include the Function return type here?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java (hunk 1 above)
Comment (nateab) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:89]: removed error, added row describing if it is a system function or not

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java (hunk 2 above)
Comment (nateab) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:114]: @AlanConfluent we could add it to the signature, or add a new row entirely, whatever makes the most sense in the future implementation, but we certainly do have that flexibility 

@pgaref Type inference is a complicated mechanism, and it is not always straightforward on how/what return type to provide for a function. Currently you would need the `CallContext` in order to infer the output type of the result.

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 3 above)
Comment (nateab) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q:360]: Yes i totally agree about it making difficult to read, but this the current standard. The variables are defined here 
https://github.com/apache/flink/blob/master/flink-table/flink-sql-client/src/test/java/org/apache/flink/table/client/cli/CliClientITCase.java#L148-L168 and this how we handle using the generated temporary jar paths in the tests, while still having correctly formatted tables

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java (hunk 4 above)
Comment (nateab) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:125]: I agree that info seems a bit redundant and informal, but this is consistent with what `DESCRIBE CATALOG` currently does https://nightlies.apache.org/flink/flink-docs-master/docs/dev/table/sql/describe/#describe-statements. I would be open to changing it, but preferably to change both to keep them consistent. Snowflake for example does `property` and `value` https://docs.snowflake.com/en/sql-reference/sql/desc-function

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 5 above)
Comment (nateab) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q:1]: why not include `SHOW`?

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 6 above)
Comment (nateab) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q:268]: removed

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeFunction.java (hunk 7 above)
Comment (nateab) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeFunction.java:68]: Intellij prefers the current way, 
Screenshot 2024-08-06 at 2 53 19 PM
and when I looked it up, it does seem to be the reccommended approach to use an empty array instead of a pre-sized array
https://stackoverflow.com/questions/53284214/toarray-with-pre-sized-array

**Comment by nateab** [comment]
@flinkbot run azure

**Comment by nateab** [comment]
@flinkbot run azure

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java (hunk 4 above)
Comment (AlanConfluent) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:119]: I see.  Ok, works for me.

**Review by AlanConfluent — APPROVED** [review]

**Review by pgaref — APPROVED** [review]
Thanks for the changes @nateab ! Latest PR LGTM

File: docs/content/docs/dev/table/sql/describe.md [hunk 8]
```diff
@@ -26,7 +26,7 @@ under the License.
 
 # DESCRIBE Statements
 
-DESCRIBE statements are used to describe the schema of a table or a view, or the metadata of a catalog, or the specified job in the Flink cluster.
+DESCRIBE statements are used to describe the schema of a table or a view, or the metadata of a catalog or a function, or the specified job in the Flink cluster.
```
Comment (jnh5y) [inline docs/content/docs/dev/table/sql/describe.md:29]: Can you add the English to https://github.com/apache/flink/blob/master/docs/content.zh/docs/dev/table/sql/describe.md as well?

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 5 above)
Comment (jnh5y) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q:1]: Is there a SHOW FUNCTION?  Or is it SHOW FUNCTIONS?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeModel.java [hunk 9]
```diff
@@ -41,7 +41,7 @@ public class SqlRichDescribeModel extends SqlCall {
     public static final SqlSpecialOperator OPERATOR =
             new SqlSpecialOperator("DESCRIBE MODEL", SqlKind.OTHER);
     protected final SqlIdentifier modelNameIdentifier;
-    private boolean isExtended;
+    private final boolean isExtended;
```
Comment (jnh5y) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeModel.java:44]: Is this change needed?

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeTable.java [hunk 10]
```diff
@@ -39,7 +39,7 @@ public class SqlRichDescribeTable extends SqlCall {
     public static final SqlSpecialOperator OPERATOR =
             new SqlSpecialOperator("DESCRIBE TABLE", SqlKind.DESCRIBE_TABLE);
     protected final SqlIdentifier tableNameIdentifier;
-    private boolean isExtended = false;
+    private final boolean isExtended;
```
Comment (jnh5y) [inline flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeTable.java:42]: Same, is this necessary?

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java [hunk 11]
```diff
@@ -448,6 +448,15 @@ void testDescribeModel() {
         sql("desc model catalog1.db1.mdl").ok("DESCRIBE MODEL `CATALOG1`.`DB1`.`MDL`");
     }
 
+    @Test
+    void testDescribeFunction() {
+        sql("describe function fn").ok("DESCRIBE FUNCTION `FN`");
+        sql("describe function catalog1.db1.fn").ok("DESCRIBE FUNCTION `CATALOG1`.`DB1`.`FN`");
+
+        sql("desc function fn").ok("DESCRIBE FUNCTION `FN`");
+        sql("desc function catalog1.db1.fn").ok("DESCRIBE FUNCTION `CATALOG1`.`DB1`.`FN`");
```
Comment (jnh5y) [inline flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:457]: Maybe add a test case for `DESCRIBE FUNCTION EXTENDED`?

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 5 above)
Comment (nateab) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q:1]: it is SHOW FUNCTIONS plural, but there are tests in this file that cover SHOW. This is also in line with the descriptions of table.q and model.q

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeModel.java (hunk 9 above)
Comment (nateab) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeModel.java:44]: not strictly needed, but the IDE preferred it this way, so i wanted to leave the codebase cleaner than I found it

File: flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeTable.java (hunk 10 above)
Comment (nateab) [reply flink-table/flink-sql-parser/src/main/java/org/apache/flink/sql/parser/dql/SqlRichDescribeTable.java:42]: not strictly needed, but the IDE preferred it this way, so i wanted to leave the codebase cleaner than I found it

File: docs/content/docs/dev/table/sql/describe.md (hunk 8 above)
Comment (nateab) [reply docs/content/docs/dev/table/sql/describe.md:29]: added

File: flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java (hunk 11 above)
Comment (nateab) [reply flink-table/flink-sql-parser/src/test/java/org/apache/flink/sql/parser/FlinkSqlParserImplTest.java:457]: added

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.apache.flink.table.api.internal.TableResultUtils.buildTableResult;
+import static org.apache.flink.table.types.inference.TypeInferenceUtil.generateSignature;
+
+/**
+ * Operation to describe a DESCRIBE [EXTENDED] [[catalogName.] dataBasesName].sqlIdentifier
+ * statement.
```
Comment (dawidwys) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:44]: Sth is off here.
Maybe sth like:
```
Operation to describe a FUNCTION.

<p>Syntax:
<pre>
DESCRIBE FUNCTION [EXTENDED] [[catalogName.] dataBasesName].sqlIdentifier
</pre>
```

**Review by dawidwys — APPROVED** [review]
Just a tiny comment to a javadoc. Do you mind rebasing this as well? After it passes CI I'll merge.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+            rows.add(
+                    Arrays.asList(
+                            "resource uris", catalogFunction.getFunctionResources().toString()));
+        }
+
+        if (isExtended) {
+            final FunctionDefinition definition = function.getDefinition();
+            rows.add(Arrays.asList("kind", definition.getKind().toString()));
+            rows.add(Arrays.asList("requirements", definition.getRequirements().toString()));
+            rows.add(Arrays.asList("deterministic", String.valueOf(definition.isDeterministic())));
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:105]: can we use `is determinstic`, `is temporary` and `supports constant folding`, `is system function`? reads nicer.

File: flink-table/flink-sql-client/src/test/resources/sql/function.q [hunk 14]
```diff
@@ -346,3 +346,159 @@ show user functions;
...
+|     resource uris | [ResourceUri{resourceType=JAR, uri='$VAR_UDF_JAR_PATH'}] |
++-------------------+---------------------------------------------$VAR_UDF_JAR_PATH_DASH+
+5 rows in set
+!ok
+
+desc function extended temp_upperudf;
++-------------------+---------------------------------------------$VAR_UDF_JAR_PATH_DASH+
+|         info name |                                 $VAR_UDF_JAR_PATH_SPACE info value |
++-------------------+---------------------------------------------$VAR_UDF_JAR_PATH_DASH+
+|   system function |                                      $VAR_UDF_JAR_PATH_SPACE false |
```
Comment (twalthr) [inline flink-table/flink-sql-client/src/test/resources/sql/function.q:431]: make order determistic i.e. sort by first column

File: flink-table/flink-sql-client/src/test/resources/sql/function.q [hunk 15]
```diff
@@ -346,3 +346,159 @@ show user functions;
...
+|        class name |                                   $VAR_UDF_JAR_PATH_SPACE UpperUDF |
+| function language |                                       $VAR_UDF_JAR_PATH_SPACE JAVA |
+|     resource uris | [ResourceUri{resourceType=JAR, uri='$VAR_UDF_JAR_PATH'}] |
++-------------------+---------------------------------------------$VAR_UDF_JAR_PATH_DASH+
+5 rows in set
+!ok
+
+desc function extended temp_upperudf;
++-------------------+---------------------------------------------$VAR_UDF_JAR_PATH_DASH+
+|         info name |                                 $VAR_UDF_JAR_PATH_SPACE info value |
```
Comment (twalthr) [inline flink-table/flink-sql-client/src/test/resources/sql/function.q:429]:
```suggestion
|         attribute name |                                 $VAR_UDF_JAR_PATH_SPACE attribute value |
```

**Review by twalthr — COMMENTED** [review]
Some final nitpicking. Otherwise +1 to this.

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 15 above)
Comment (nateab) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q:429]: This would deviate from `DESCRIBE CATALOG` currently does, are you okay with that? Also see my prior comment on the naming here https://github.com/apache/flink/pull/25115#discussion_r1706084728

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 14 above)
Comment (nateab) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q:431]: is that necessary? we don't do that for `DESCRIBE CATALOG` for example. I tried to order the rows in a way that I thought was logical

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java (hunk 12 above)
Comment (nateab) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:44]: added

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java (hunk 13 above)
Comment (nateab) [reply flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/DescribeFunctionOperation.java:105]: added

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 14 above)
Comment (dawidwys) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q]: I agree with @nateab, if required let's change both for `DESCRBIE CATALOG` and `FUNCTION` in a follow up.

File: flink-table/flink-sql-client/src/test/resources/sql/function.q (hunk 15 above)
Comment (dawidwys) [reply flink-table/flink-sql-client/src/test/resources/sql/function.q]: I agree with @nateab, if required let's change both for `DESCRBIE CATALOG` and `FUNCTION` in a follow up.
